fhirflat/ingest.py

```diff
@@ -0,0 +1,279 @@
+                    else:
+                        result[key] = [result[key], snippet[key]]
+    return result
+
+
+def create_dict_from_cell(row, full_df, map_df):
+    """
+    Iterates through the columns of the row, applying the mapping to each columns
+    and produces a fhirflat-like dictionary to initialize the resource object.
+    """
+
+    column = row["column"]
+    response = row["value"]
+    if pd.notna(response):  # Ensure there is a response to map
+        try:
+            # Retrieve the mapping for the given column and response
+            if pd.isna(map_df.loc[column].index).all():
+                mapping = map_df.loc[(column, np.nan)].dropna()
+            else:
+                mapping = map_df.loc[(column, str(int(response)))].dropna()
+            snippet = {
+                k: (
+                    v
+                    if "<" not in str(v)
+                    else find_field_value(row, response, v, raw_data=full_df)
+                )
+                for k, v in mapping.items()
+            }
+            return snippet
+        except KeyError:
+            # No mapping found for this column and response despite presence
+            # in mapping file
+            warnings.warn(
+                f"No mapping for column {column} response {response}",
+                UserWarning,
+            )
+            return None
+
+
+def create_dictionary(
+    data: pd.DataFrame, map_file: pd.DataFrame, resource: str, one_to_one=False
+) -> pd.DataFrame:
+    """
+    Given a data file and a single mapping file for one FHIR resource type,
+    returns a single column dataframe with the mapped data in a FHIRflat-like
+    format, ready for further processing.
+
+    Parameters
+    ----------
+    data: pd.DataFrame
+        The data file containing the clinical data.
+    map_file: pd.DataFrame
+        The mapping file containing the mapping of the clinical data to the FHIR
+        resource.
+    resource: str
+        The name of the resource being mapped.
+    one_to_one: bool
+        Whether the resource should be mapped as one-to-one or one-to-many.
+    """
+
+    data = pd.read_csv(data, header=0)
+    map_df = pd.read_csv(map_file, header=0)
+
+    # setup the data -----------------------------------------------------------
+    relevant_cols = map_df["redcap_variable"].dropna().unique()
+    filtered_data = data.loc[:, data.columns.isin(relevant_cols)].copy()
+
+    if filtered_data.empty:
+        warnings.warn(f"No data found for the {resource} resource.", UserWarning)
+        return None
+
+    if not one_to_one:
+        filtered_data = filtered_data.reset_index()
+        melted_data = filtered_data.melt(id_vars="index", var_name="column")
+
+    # set up the mappings -------------------------------------------------------
+
+    # Fills the na redcap variables with the previous value
+    map_df["redcap_variable"] = map_df["redcap_variable"].ffill()
```

Comment:
Is the transformation file specific to redcap?

Reply:
Edited to "raw_variable" and "raw_response" to make more generic